platform/bukkit/plugin/src/main/java/me/lokka30/treasury/plugin/bukkit/hooks/EconomyHook.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * This file is/was part of Treasury. To read more information about Treasury such as its licensing, see <https://github.com/lokka30/Treasury>.
+ */
+
+package me.lokka30.treasury.plugin.bukkit.hooks;
+
+import java.math.BigDecimal;
+import java.text.DecimalFormat;
+import java.util.Locale;
+import java.util.Optional;
+import me.lokka30.treasury.api.common.service.Service;
+import me.lokka30.treasury.api.common.service.ServiceRegistry;
+import me.lokka30.treasury.api.economy.EconomyProvider;
+import me.lokka30.treasury.api.economy.account.PlayerAccount;
+import me.lokka30.treasury.api.economy.currency.Currency;
+import me.lokka30.treasury.api.economy.response.EconomySubscriber;
+import me.lokka30.treasury.plugin.bukkit.TreasuryBukkit;
+import org.bukkit.OfflinePlayer;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class EconomyHook implements TreasuryPAPIHook {
+
+    private EconomyProvider provider;
+    private final DecimalFormat format = new DecimalFormat("#,###");
+    private final PAPIExpansion expansion;
+    private final TreasuryBukkit plugin;
+    private final String k, m, b, t, q;
+    private Baltop baltop;
+
+    public EconomyHook(PAPIExpansion expansion, TreasuryBukkit plugin) {
+        this.expansion = expansion;
+        this.plugin = plugin;
+        this.k = expansion.getString("formatting.thousands", "k");
+        this.m = expansion.getString("formatting.millions", "m");
+        this.b = expansion.getString("formatting.billions", "b");
+        this.t = expansion.getString("formatting.trillions", "t");
+        this.q = expansion.getString("formatting.quadrillions", "q");
+    }
+
+    @Override
+    public boolean setup() {
+        Optional<Service<EconomyProvider>> serviceOpt = ServiceRegistry.INSTANCE.serviceFor(
+                EconomyProvider.class);
+        if (serviceOpt.isPresent()) {
+            provider = serviceOpt.get().get();
+            this.baltop = new Baltop(
+                    expansion.getBoolean("baltop.enabled", false),
+                    expansion.getInt("baltop.cache_size", 100),
+                    expansion.getInt("baltop.cache_delay", 60),
+                    provider
+            );
+            this.baltop.start(plugin);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean canRegister() {
+        return provider != null;
+    }
+
+    @Override
+    public @Nullable String onRequest(
+            @Nullable OfflinePlayer player, @NotNull String param
+    ) {
+        if (provider != null) {
+            if (!baltop.isEnabled() && param.startsWith("top_")) {
+                return (param.startsWith("top_balance")) ? "0" : "";
+            }
+
+            if (param.contains("top_balance")) {
+                if (param.startsWith("top_balance_fixed")) {
+                    String currencyId = param.replace("top_balance_fixed_", "");
```

Comment:
This results in a runaround for `top_balance_fixed` where a regular expression would not. Same for all other entries that accept parameters but do not require them.

Reply:
i have no idea on how to use patterns.